src/utils/tabs.rs

```diff
@@ -0,0 +1,74 @@
+use crate::config::SPACES;
+
+use unicode_segmentation::UnicodeSegmentation;
+
+pub fn clamp_arg(tab_size: usize) -> usize {
+    std::cmp::min(tab_size, SPACES.len())
+}
+
+/// Expand tabs as spaces.
+/// tab_width = 0 is documented to mean do not replace tabs.
+pub fn expand(line: &str, tab_width: usize) -> String {
+    let line_bytes = line.as_bytes();
+    if tab_width == 0 || !line_bytes.iter().any(|c| *c == b'\t') {
+        line.to_string()
+    } else {
+        let expanded = Vec::with_capacity(line.len());
+
+        let expand_tabs = |mut v: Vec<u8>, c: &u8| -> Vec<u8> {
+            if *c == b'\t' {
+                v.extend_from_slice(SPACES[..tab_width].as_bytes());
+            } else {
+                v.push(*c);
+            }
+
+            v
+        };
+        let expanded = line_bytes.iter().fold(expanded, expand_tabs);
+
+        // Safety: when expanding ascii '\t' to space(s) a buffer will remain valid utf-8.
+        std::str::from_utf8(&expanded)
+            .unwrap_or_else(|err| {
+                panic!(
+                    "Internal error: invalid utf-8 when expanding tabs of line {:?}: {}",
+                    line, err
+                )
+            })
+            .to_string()
+    }
+}
+
+/// Remove `prefix` chars from `line`, then call `tabs::expand()`.
+pub fn remove_prefix_and_expand(prefix: usize, line: &str, tab_width: usize) -> String {
+    let line_bytes = line.as_bytes();
+    // The to-be-removed prefixes are almost always ascii +/- (or ++/ +/.. for merges) for
+    // which grapheme clusters are not required.
+    if line_bytes.len() >= prefix && line_bytes[..prefix].is_ascii() {
+        // Safety: slicing into the utf-8 line-str is ok, upto `prefix` only ascii was present
+        expand(&line[prefix..], tab_width)
+    } else {
+        let cut_line = line.graphemes(true).skip(prefix).collect::<String>();
+        expand(&cut_line, tab_width)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+
+    #[test]
+    fn test_remove_prefix_and_expand() {
```

Comment:
Can you add a test that try to expand a utf8 char that has the tab ascii code in it?

like this for example:
`क`

Reply:
Well, that's the beauty of the utf-8 representation of unicode codepoints, all non-ascii bytes have the highest/0x80 bit set and a tab does not, so such a utf-8 char does not exist (unlike in utf-16).